core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -417,6 +417,33 @@ public void reconfigure(ServerConfigurator serverConfigurator) {
         pipelineConfigurator.updateConfig(config);
     }
 
+    /**
+     * Add a shutdown hook to stop this {@link Server}.
+     */
+    public void closeOnShutdown() {
+        closeOnShutdown(new Thread(() -> {
+            stop().join();
+            logger.info("Server has been stopped.");
+        }));
+    }
+
+    /**
+     * Add a shutdown hook to run a given {@link Runnable}. You need to call {@link Server#stop()}
+     * in a runnable to stop this {@link Server}.
+     *
+     * <pre>{@code
+     * server.closeOnShutDown(() -> {
+     *    // Do something before stopping this server.
+     *    server.stop().join();
+     *    // Do something after stopping this server.
+     * })
+     * }</pre>
+     */
+    public void closeOnShutdown(Runnable runnable) {
```

Comment:
How about changing the return type of this method instead? e.g.

```java
public CompletableFuture<Void> closeOnShutdown() { ... }

server.closeOnShutdown().thenRun(() -> {
    System.err.println("Stopped");
}).exceptionally(cause -> {
    cause.printStackTrace();
});
```

We could also make the similar change to `ClientFactory.closeOnShutdown()`.

Reply:
In this case, can users run something before a server/client is shutdown?

Reply:
I think it would be nice to call `server.stop()` or `server.close()` automatically.
`server.close()` permantely stop a `Server`. Is it better to call `server.closeAsync()` because a JVM is shutting down?

> In this case, can users run something before a server/client is shutdown?
How about adding a variant that takes a `Runnable` to execute before closing?
```java
class Server {

    CompletableFuture<Void> closeOnShutdown() {
        return closeOnShutdown(null);
    }

    CompletableFuture<Void> closeOnShutdown(@Nullable Runnable whenClosing) {
        ...
    }
}
```